services/web/server/src/simcore_service_webserver/garbage_collector_tasks_users.py

```diff
@@ -0,0 +1,109 @@
+"""
+    Scheduled tasks addressing users
+
+"""
+
+import asyncio
+import logging
+from typing import AsyncIterator, Callable
+
+from aiohttp import web
+from aiopg.sa.engine import Engine
+from models_library.users import UserID
+from servicelib.logging_utils import log_context
+from tenacity import retry
+from tenacity.before_sleep import before_sleep_log
+from tenacity.wait import wait_exponential
+
+from ._constants import APP_DB_ENGINE_KEY
+from .login.utils import notify_user_logout
+from .security_api import clean_auth_policy_cache
+from .users_db import update_expired_users
+
+logger = logging.getLogger(__name__)
+
+CleanupContextFunc = Callable[[web.Application], AsyncIterator[None]]
+
+_SEC = 1
+
+_PERIODIC_TASK_NAME = f"{__name__}.update_expired_users_periodically"
+_APP_TASK_KEY = f"{_PERIODIC_TASK_NAME}.task"
+
+
+async def notify_user_logout_all_sessions(
+    app: web.Application, user_id: UserID
+) -> None:
+    #  NOTE kept here for convenience
+    with log_context(
+        logger,
+        logging.INFO,
+        "Forcing logout of %s from all sessions",
+        f"{user_id=}",
+    ):
+        try:
+            await notify_user_logout(app, user_id, client_session_id=None)
+        except Exception:  # pylint: disable=broad-except
+            logger.warning(
+                "Ignored error while notifying logout for %s",
+                f"{user_id=}",
+                exec_info=True,
+            )
+
+
+@retry(
+    wait=wait_exponential(min=5 * _SEC),
+    before_sleep=before_sleep_log(logger, logging.WARNING),
```

Comment:
Yes, the intention is to be resilient. When it raises ``Exception`` , it should restart after a pause. For instance, if the database disconnects or is restarted etc ...

Now regarding the wait pattern ... i am not sure what do you mean by "predictable". As far as I recall, you explained to me that when it comes to connection failures with other services (e.g. db) the best is to do exponential to you give enough time to the service to recover in case it is saturated of "insisting clients". What is the difference here? 
I do not have a strong opinion on one or the other, i simply not sure why one is better than the other or what you mean by predicatble?

Reply:
I see the confusion here.

The functions where you want to apply the exponential retry policy are `update_expired_users` and `clean_auth_policy_cache`. Actually what you want to do is apply a retry block on the code from line `68 => 84` since it includes all your sensible code. I would wrap that in a function with the above retry policy. This also requires a try catch block. I think this pattern is used in most similar places. (at some point we might want to wrap it and create a standard out of it).

I would refactor it like below. Please adjust for additional logging if required.

```python
@retry(
    wait=wait_exponential(min=5 * _SEC),
    before_sleep=before_sleep_log(logger, logging.WARNING),
)
async def __clean_expired_users(app: web.Application, engine: Engine) -> None:
    if updated := await update_expired_users(engine):
        # expired users might be cached in the auth. If so, any request
        # with this user-id will get thru producing unexpected side-effects
        clean_auth_policy_cache(app)

        # broadcast force logout of user_id
        for user_id in updated:
            logger.info("User account with %s expired", f"{user_id=}")

            # NOTE: : this notification will never reach sockets because it runs in the GC!!
            # We need a mechanism to send messages from GC to the webservers
            # OR a way to notify from the database changes back to the web-servers (similar to compuational services)
            # SEE https://github.com/ITISFoundation/osparc-simcore/issues/3387
            # await notify_user_logout_all_sessions(app, user_id)

    else:
        logger.info("No users expired")


async def _update_expired_users_periodically(
    app: web.Application, wait_interval_s: float
):
    """Periodically check expiration dates and updates user status
    It is resilient, i.e. if update goes wrong, it waits a bit and retries
    """
    engine: Engine = app[APP_DB_ENGINE_KEY]
    assert engine  # nosec

    while True:
        try:
            await __clean_expired_users(app, engine)
        except Exception:
            pass  # log if required or just skip like this
        await asyncio.sleep(wait_interval_s)
```

Reply:
done.